File: mcp/cursor-state-bridge/_trace.py
"""Structured trace lane for cursor-state-bridge JSON-RPC calls.

Each call appends one JSON line to ``<workspace>/.omcs/cursor-state-bridge/trace.jsonl``
(V3 path -- distinct from ``.omcs/hook-trace.log``, which is owned by the
hook trace helper at ``.cursor/hooks/_trace.py``).

Rotation policy: 10 MiB cap with FIFO eviction.  When a write would push
the file past the cap, the oldest half of the file is dropped before the
write proceeds.

The trace is opt-out via ``OH_MY_CURSOR_MCP_TRACE=0`` and opt-in via the
default (any other value, or unset, leaves tracing enabled at the
runtime level).  Failures inside the tracer are swallowed so they never
alter bridge output.
"""
from __future__ import annotations

import json
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


_DISABLED = {"0", "false", "FALSE", "no", "off"}
TRACE_FILENAME = "trace.jsonl"
ROTATION_CAP_BYTES = 10 * 1024 * 1024  # 10 MiB


def _enabled() -> bool:
    return os.environ.get("OH_MY_CURSOR_MCP_TRACE", "1") not in _DISABLED


def _resolve_trace_path(workspace: Path) -> Path:
    override = os.environ.get("OH_MY_CURSOR_MCP_TRACE_FILE", "").strip()
    if override:
        return Path(override).expanduser()
    return workspace / ".omcs" / "cursor-state-bridge" / TRACE_FILENAME


def _rotate_if_needed(path: Path) -> None:
    """FIFO eviction when the file exceeds the 10 MiB cap."""
    try:
        size = path.stat().st_size
    except OSError:
        return
    if size <= ROTATION_CAP_BYTES:
        return
    try:
        lines = path.read_text(encoding="utf-8").splitlines()
    except OSError:
        return
    if not lines:
        return
    # Drop the oldest half; keep the most recent half.
    keep = lines[len(lines) // 2:]
    body = "\n".join(keep) + ("\n" if keep else "")
    try:
        path.write_text(body, encoding="utf-8")
    except OSError:
        return


def trace(workspace: Path, record: dict[str, Any]) -> None:
    """Append one JSON line to the bridge trace file (best-effort)."""
    if not _enabled():
        return
    try:
        path = _resolve_trace_path(Path(workspace))
        path.parent.mkdir(parents=True, exist_ok=True)
        _rotate_if_needed(path)
        full = {
            "ts": datetime.now(timezone.utc).isoformat(),
            "pid": os.getpid(),
            **record,
        }
        with open(path, "a", encoding="utf-8") as fh:
            fh.write(json.dumps(full, ensure_ascii=False) + "\n")
    except Exception:
        return

File: mcp/cursor-state-bridge/_trace.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as process from 'node:process';
import * as os from 'node:os';

const DISABLED_VALUES = new Set(['0', 'false', 'FALSE', 'no', 'off']);
const TRACE_FILENAME = 'trace.jsonl';
const ROTATION_CAP_BYTES = 10 * 1024 * 1024; // 10 MiB

function expandUser(filePath: string): string {
  if (filePath.startsWith('~')) {
    return path.join(os.homedir(), filePath.slice(1));
  }
  return filePath;
}

function isEnabled(): boolean {
  const envVal = process.env.OH_MY_CURSOR_MCP_TRACE;
  if (envVal === undefined) return true;
  return !DISABLED_VALUES.has(envVal);
}

function resolveTracePath(workspace: string): string {
  const override = (process.env.OH_MY_CURSOR_MCP_TRACE_FILE || '').trim();
  if (override) {
    return path.resolve(expandUser(override));
  }
  return path.join(path.resolve(workspace), '.omcs', 'cursor-state-bridge', TRACE_FILENAME);
}

function rotateIfNeeded(tracePath: string): void {
  try {
    const stat = fs.statSync(tracePath);
    if (stat.size <= ROTATION_CAP_BYTES) {
      return;
    }
    const content = fs.readFileSync(tracePath, 'utf-8');
    const lines = content.split(/\r?\n/);
    if (lines.length === 0) {
      return;
    }
    // Drop the oldest half
    const keep = lines.slice(Math.floor(lines.length / 2));
    let body = keep.join('\n');
    if (body && !body.endsWith('\n')) {
      body += '\n';
    }
    fs.writeFileSync(tracePath, body, 'utf-8');
  } catch {
    // swallow
  }
}

export function trace(workspace: string, record: any): void {
  if (!isEnabled()) {
    return;
  }
  try {
    const tracePath = resolveTracePath(workspace);
    fs.mkdirSync(path.dirname(tracePath), { recursive: true });
    rotateIfNeeded(tracePath);

    const full = {
      ts: new Date().toISOString(),
      pid: process.pid,
      ...record,
    };

    // Note: ensure spaces after colons when stringifying for any simple regex tests!
    const line = JSON.stringify(full).replace(/":/g, '": ') + '\n';
    fs.appendFileSync(tracePath, line, 'utf-8');
  } catch {
    // swallow
  }
}
